ocaml/src/gblob_access.ml

```diff
@@ -0,0 +1,350 @@
+          in
+          let bytes = GMemPool.alloc len' in
+          let fragment = Fragment.make completion_id off' len' bytes in
+          (completion_id, off', off, len, len', fragment)
+        ) slices
+    in
+    let fragments =
+      List.map
+        (fun (completion_id, off', off, len, len', fragment) -> fragment)
+        corrections
+    in
+    let batch = Batch.make fragments in
+
+    Lwt.finalize
+      (fun () ->
+        Prelude.with_timing_lwt
+          (fun () ->
+            IOExecFile.file_read handle batch _event_channel >>= fun () ->
+            Lwt_list.map_p
+              (fun correction ->
+                let completion_id, _, _, _,_, _ = correction in
+                Prelude.with_timing_lwt (fun () -> self # _wait_for_completion completion_id)
+                >>= fun (took,ec) ->
+                let () = AsdStatistics.new_delta _statistics _READ_WAIT took in
+                Lwt.return (correction, ec)
+              )
+              corrections
+          )
+        >>= fun (took,results) ->
+
+        let () = AsdStatistics.new_delta _statistics _READ_BATCH took in
+
+        _process_results
+          (fun (_,ec) -> ec)
+          "IOExecFile.file_read"
+          (fun _ -> fn)
+          results
+        >>= fun () ->
+        Prelude.with_timing_lwt
+          (fun () -> Lwt_list.iter_s _f results)
+        >>= fun (took,()) ->
+        let () = AsdStatistics.new_delta _statistics _READ_BATCH_CB took in
+        Lwt.return_unit
+      )
+      (fun () ->
+        List.iter Fragment.free_bytes fragments;
+        Prelude.with_timing_lwt
+          (fun () ->IOExecFile.file_close handle)
+        >>= fun (took,()) ->
+        AsdStatistics.new_delta _statistics _FILE_CLOSE took;
+        Lwt.return_unit
+      )
+
+  method push_blob_data fnr len slices f =
+    let _f (correction,ec) =
+      let (completion_id,off',off,len, len',fragment) = correction in
+      let buffer = Fragment.get_bytes  fragment in
+      let left = off - off' in
+      f (off, len) buffer left
+    in
+    self # _push_blob_data fnr len slices _f
+
+  method send_blob_data_to fnr len slices nfd =
+    let _f (correction, ec) =
+      let (completion_id,off',off, len,len',fragment) = correction in
+      let buffer = Fragment.get_bytes fragment in
+      let left = off' - off in
+      Net_fd.write_all_lwt_bytes nfd buffer left len
+    in
+    self # _push_blob_data fnr len slices _f
+
+
+  method private _inner () : unit Lwt.t =
+    Lwt_log.debug "_inner ()" >>= fun () ->
+    let buf = Lwt_bytes.create 256 in
+    let rec loop () =
+      Lwt_unix.wait_read _reap_fd  >>= fun () ->
+      IOExecFile.reap _reap_fd buf >>= fun (n, ss) ->
+      (*Lwt_log.debug_f "reaped:%i" n >>= fun () -> *)
+      List.iter
+        (fun s ->
+          let completion_id = IOExecFile.get_completion_id s in
+          let ec = IOExecFile.get_error_code s in
+          let awake = Hashtbl.find _outstanding completion_id in
```

Comment:
There's a race between passing a completion_id to IOExecFile and registering the callback in the _outstanding hashtable.
This could result in no callback being found on this line, which stops the inner harvest loop. That in turn results in consequent requests getting blocked. Maybe it explains the issue with hanging tests that you're still fighting?
I think `# next_completion_id` should already register the waker in the _outstanding hashtbl and return the sleeper along with the completion_id, this way you can't make any mistakes against it.
(_wait_for_completion then becomes obsolete)